File: typescript/src/ir.ts
// JSON IR types and loader. Mirrors specs/parser-spec.md.
import { readFileSync } from "node:fs";

export interface Step {
  keyword: string;
  text: string;
  parameters: string[];
}

export interface Scenario {
  name: string;
  steps: Step[];
  examples: Record<string, string>[];
}

export interface Feature {
  name: string;
  background: Step[];
  scenarios: Scenario[];
}

function normalizeStep(raw: unknown): Step {
  const s = raw as { keyword?: string; text?: string; parameters?: string[] };
  if (typeof s.keyword !== "string" || typeof s.text !== "string") {
    throw new Error("step IR missing keyword or text");
  }
  return {
    keyword: s.keyword,
    text: s.text,
    parameters: Array.isArray(s.parameters) ? [...s.parameters] : [],
  };
}

function normalizeScenario(raw: unknown): Scenario {
  const s = raw as {
    name?: string;
    steps?: unknown[];
    examples?: Record<string, string>[];
  };
  if (typeof s.name !== "string") {
    throw new Error("scenario IR missing name");
  }
  return {
    name: s.name,
    steps: (s.steps ?? []).map(normalizeStep),
    examples: (s.examples ?? []).map((row) => ({ ...row })),
  };
}

export function fromObject(raw: unknown): Feature {
  const r = raw as {
    name?: string;
    background?: unknown[];
    scenarios?: unknown[];
  };
  if (typeof r.name !== "string") {
    throw new Error("feature IR missing 'name'");
  }
  return {
    name: r.name,
    background: (r.background ?? []).map(normalizeStep),
    scenarios: (r.scenarios ?? []).map(normalizeScenario),
  };
}

export function loadIR(path: string): Feature {
  return fromObject(JSON.parse(readFileSync(path, "utf-8")));
}

File: python/src/aps_kit/ir.py
"""JSON IR types and loader. Mirrors specs/parser-spec.md."""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


@dataclass(frozen=True)
class Step:
    keyword: str
    text: str
    parameters: tuple[str, ...] = ()

    @classmethod
    def from_obj(cls, obj: dict[str, Any]) -> "Step":
        return cls(
            keyword=obj["keyword"],
            text=obj["text"],
            parameters=tuple(obj.get("parameters") or ()),
        )


@dataclass(frozen=True)
class Scenario:
    name: str
    steps: tuple[Step, ...]
    examples: tuple[dict[str, str], ...]

    @classmethod
    def from_obj(cls, obj: dict[str, Any]) -> "Scenario":
        return cls(
            name=obj["name"],
            steps=tuple(Step.from_obj(s) for s in obj.get("steps", [])),
            examples=tuple(obj.get("examples") or ()),
        )


@dataclass(frozen=True)
class Feature:
    name: str
    background: tuple[Step, ...] = ()
    scenarios: tuple[Scenario, ...] = field(default_factory=tuple)

    @classmethod
    def from_obj(cls, obj: dict[str, Any]) -> "Feature":
        if "name" not in obj:
            raise ValueError("feature IR missing 'name'")
        return cls(
            name=obj["name"],
            background=tuple(Step.from_obj(s) for s in obj.get("background") or ()),
            scenarios=tuple(Scenario.from_obj(s) for s in obj.get("scenarios") or ()),
        )


def load_ir(path: str | Path) -> Feature:
    with open(path, encoding="utf-8") as fh:
        return Feature.from_obj(json.load(fh))
